File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static boolean[] visited;
    static TreeSet<Integer>[] arr;
    static StringBuilder sb;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        sb = new StringBuilder();
        N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        arr = new TreeSet[N + 1];
        visited = new boolean[N + 1];
        for(int i = 1; i <= N; i++) {
            arr[i] = new TreeSet<>();
        }
        for(int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            arr[a].add(b);
            arr[b].add(a);
        }
        // DFS
        DFS(K);
        System.out.println(sb.toString().trim());
        // BFS
        sb = new StringBuilder();
        visited = new boolean[N + 1];
        Queue<Integer> que = new LinkedList<>();
        que.add(K);
        visited[K] = true;
        while(!que.isEmpty()) {
            int x = que.poll();
            sb.append(x).append(" ");
            for(int num : arr[x]) {
                if(!visited[num]) {
                    que.add(num);
                    visited[num] = true;
                }
            }
        }
        System.out.println(sb.toString().trim());
    }

    static void DFS(int node) {
        visited[node] = true;
        sb.append(node).append(" ");
        for(int x : arr[node]) {
            if(!visited[x]) {
                DFS(x);
            }
        }
    }
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js
class Queue {
  constructor() {
    this.arr = [];
    this.head = 0;
  }

  add(num) {
    this.arr.push(num);
  }

  poll() {
    return this.arr[this.head++];
  }

  isEmpty() {
    return this.arr.length === this.head;
  }
}

const [[N, M, V], ...input] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n').map(line => line.trim().split(' ').map(Number));
const arr = Array.from({length : N + 1}, () => []);
let visited = new Array(N + 1).fill(false);
for(let i = 0; i < M; i++) {
  arr[input[i][0]].push(input[i][1]);
  arr[input[i][1]].push(input[i][0]);
}
for (let i = 1; i <= N; i++) arr[i].sort((a, b) => a - b);
const ans = [[], []];
DFS(V);
const que = new Queue();
visited = visited.fill(false);
que.add(V);
visited[V] = true;
while(!que.isEmpty()) {
  const node = que.poll();
  ans[1].push(node);
  for(let num of arr[node]) {
    if(!visited[num]) {
      que.add(num);
      visited[num] = true;
    }
  }
}
console.log(ans.map(line => line.join(' ')).join('\n'));
function DFS(x) {
  ans[0].push(x);
  visited[x] = true;
  for(let num of arr[x]) {
    if(!visited[num]) {
      DFS(num);
    }
  }
}